### ISSUE TITLE
[Bug]: Milvus crashes during multi-threaded write with long partition names

### ISSUE THREAD

**Issue by AlexeyIvanov8** [milvus-io/milvus#19596]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 2.1.2
- Deployment mode(standalone or cluster): reproduce in both modes standalone and cluster
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus==2.1.2
- OS(Ubuntu or CentOS): docker-compose from official site launched on OSX Monterey
- CPU/Memory: Intel core I7 2.6GHz, 6 cores/16Gb ram
```

### Current Behavior

When records are inserted in multiple threads into partitions with long names that created as needed, an error occurs:
<details>
<summary>error</summary>
`
milvus_1  | [2022/09/30 07:58:17.378 +00:00] [ERROR] [proxy/channels_mgr.go:103] ["failed to describe collection"] [error_code=UnexpectedError] [reason="DescribeCollection failed: can't find collection id : 0"] [stack="github.com/milvus-io/milvus/internal/proxy.getDmlChannelsFunc.func1\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/channels_mgr.go:103\ngithub.com/milvus-io/milvus/internal/proxy.(*singleTypeChannelsMgr).getPChans\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/channels_mgr.go:144\ngithub.com/milvus-io/milvus/internal/proxy.(*singleTypeChannelsMgr).getChannels\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/channels_mgr.go:164\ngithub.com/milvus-io/milvus/internal/proxy.(*channelsMgrImpl).getChannels\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/channels_mgr.go:327\ngithub.com/milvus-io/milvus/internal/proxy.(*insertTask).getChannels\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/task_insert.go:98\ngithub.com/milvus-io/milvus/internal/proxy.(*dmTaskQueue).popPChanStats\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/task_scheduler.go:301\ngithub.com/milvus-io/milvus/internal/proxy.(*dmTaskQueue).PopActiveTask\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/task_scheduler.go:258\ngithub.com/milvus-io/milvus/internal/proxy.(*taskScheduler).processTask.func1\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/task_scheduler.go:448\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:965\nruntime.panicmem\n\t/usr/local/go/src/runtime/panic.go:212\nruntime.sigpanic\n\t/usr/local/go/src/runtime/signal_unix.go:734\ngithub.com/milvus-io/milvus/internal/proxy.(*insertTask).checkLengthOfFieldsData\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/task_insert.go:107\ngithub.com/milvus-io/milvus/internal/proxy.(*insertTask).checkPrimaryFieldData\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/task_insert.go:127\ngithub.com/milvus-io/milvus/internal/proxy.(*insertTask).PreExecute\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/task_insert.go:228\ngithub.com/milvus-io/milvus/internal/proxy.(*taskScheduler).processTask\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/task_scheduler.go:452"]
milvus_1  | panic: runtime error: invalid memory address or nil pointer dereference
milvus_1  | [signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x211992d]
milvus_1  | 
milvus_1  | goroutine 294784 [running]:
milvus_1  | github.com/milvus-io/milvus/internal/proxy.(*insertTask).checkLengthOfFieldsData(0xc00104cd20, 0x203009, 0x203009)
milvus_1  | 	/go/src/github.com/milvus-io/milvus/internal/proxy/task_insert.go:107 +0x2d
milvus_1  | github.com/milvus-io/milvus/internal/proxy.(*insertTask).checkPrimaryFieldData(0xc00104cd20, 0x1, 0x1)
milvus_1  | 	/go/src/github.com/milvus-io/milvus/internal/proxy/task_insert.go:127 +0x72
milvus_1  | github.com/milvus-io/milvus/internal/proxy.(*insertTask).PreExecute(0xc00104cd20, 0x43872d8, 0xc007f34b10, 0x0, 0x0)
milvus_1  | 	/go/src/github.com/milvus-io/milvus/internal/proxy/task_insert.go:228 +0xda6
milvus_1  | github.com/milvus-io/milvus/internal/proxy.(*taskScheduler).processTask(0xc0002b1cc0, 0x43adc88, 0xc00104cd20, 0x43a9ee0, 0xc001196f90)
milvus_1  | 	/go/src/github.com/milvus-io/milvus/internal/proxy/task_scheduler.go:452 +0x56b
milvus_1  | created by github.com/milvus-io/milvus/internal/proxy.(*taskScheduler).manipulationLoop
milvus_1  | 	/go/src/github.com/milvus-io/milvus/internal/proxy/task_scheduler.go:509 +0x177
`
</details>

Error depends on number of thread(50 in example) and length of partition name(53 in my example).

### Expected Behavior

Records are written from multiple threads without errors with any partitions that could be created.

### Steps To Reproduce

```markdown
#Jupyter notebook with full example for reproduce:

# install dependencies
!pip install pymilvus==2.1.2
!pip install protobuf==3.20.0
!pip install grpcio-tools

# prepare collection
from pymilvus import connections
conn = connections.connect(
  alias="default", 
  host='host.docker.internal', 
  port='19530'
)

from pymilvus import CollectionSchema, FieldSchema, DataType
id = FieldSchema(
    name="id",
    dtype=DataType.INT64,
    is_primary=True,
    auto_id=True
)
book_name = FieldSchema(
    name="book_name", 
    dtype=DataType.VARCHAR, 
    max_length=200,
)
word_count = FieldSchema(
    name="word_count", 
    dtype=DataType.INT64,  
)
book_intro = FieldSchema(
    name="book_intro", 
    dtype=DataType.FLOAT_VECTOR, 
    dim=5
)
schema = CollectionSchema(
    fields=[id, book_name, word_count, book_intro], 
    description="Test book search"
)
collection_name = "book"

from pymilvus import Collection
collection = Collection(
    name=collection_name, 
    schema=schema, 
    using='default', 
    shards_num=20,
)

# insert partitioned data
from pymilvus import Collection
def create_partition_if_not_exists(collection, partition_name):
    if collection.has_partition(partition_name) != True:
        collection.create_partition(partition_name)
    
import random
def insert_test_data(collection):
    partition_name = 'partition_2022_0' + str(random.randint(0, 2000)) + '_03T00_00_00Z_PT1H_' + str(random.randint(1000000, 9999999)) + 'qwertyuiopa'
    data = [
      [str(100)],
      [random.randint(0, 1000)],
      [[random.random() for _ in range(5)]]
    ]
    create_partition_if_not_exists(collection, partition_name)
    mr = collection.insert(data = data, partition_name = partition_name)
 
def insert_part_of_data(max):
    collection = Collection("book")
    for i in range(max):
        insert_test_data(collection)
    return "inserted " + str(max) + " values"

from threading import Thread

# create thread for insert data
class InsertTestDataWorker(Thread):
    def __init__(self, count):
        Thread.__init__(self)
        self.count = count

    def run(self):
        insert_part_of_data(self.count)
    
def run_func(num_threads, count):
    for x in range(num_threads):
        inserter = InsertTestDataWorker(count)
        inserter.start()

# error occurs here
run_func(50, 30)
```

**Comment by xiaofan-luan** [comment]
So far milvus only support at most 4096 partitions.
And you will generate number of shards * number of partition segments, which will have great memory pressure to the system.

Questions:
What is the scenario you want to create 20 shards? usually we would say 2-4 will be well enough.
when you use partitions, are you looking for searching less data or what? we are trying to improve current behaviour but we need more proof

**Comment by AlexeyIvanov8** [comment]
@xiaofan-luan 

Use case is splitting data into time intervals to limit the data set to be searched. For example splitting data by hours.

I also retested with one shard and got the same error even though `30*50=1500` is less than the limit of 4096. But if I reduce the length of the partition name (for example `'part_' + str(random.randint(0, 4096))`) and run insert 1500 records in the same stream, then the error does not occur.

**Comment by AlexeyIvanov8** [comment]
/reopen

Retest on Milvus 2.2.6 standalone and cluster: the error is reproduced.

**Comment by yanliang567** [comment]
/assign @NicoYuan1986 
could you please help to reproduce the issue on latest milvus 2.2

**Comment by NicoYuan1986** [comment]
Reproduced.
milvus: v2.2.6
pymilvus: 2.2.6
Deployment mode: standalone
install: wget https://github.com/milvus-io/milvus/releases/download/v2.2.6/milvus-standalone-docker-compose.yml -O docker-compose.yml
cu/memory: 6 CPUs 16 GB Memory

error message:
```
>>> run_func(50, 30)
>>> 
>>> RPC error: [batch_insert], <MilvusException: (code=1, message=can't find collection name:book)>, <Time:{'RPC start': '2023-06-06 11:00:11.295063', 'RPC error': '2023-06-06 11:00:11.337184'}>
Exception in thread Thread-40:
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "<stdin>", line 6, in run
  File "<stdin>", line 4, in insert_part_of_data
  File "<stdin>", line 10, in insert_test_data
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pymilvus/orm/collection.py", line 430, in insert
    res = conn.batch_insert(self._name, entities, partition_name,
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pymilvus/decorators.py", line 109, in handler
    raise e
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pymilvus/decorators.py", line 105, in handler
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pymilvus/decorators.py", line 136, in handler
    ret = func(self, *args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pymilvus/decorators.py", line 85, in handler
    raise e
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pymilvus/decorators.py", line 50, in handler
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pymilvus/client/grpc_handler.py", line 399, in batch_insert
    raise err
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pymilvus/client/grpc_handler.py", line 395, in batch_insert
    raise MilvusException(response.status.error_code, response.status.reason)
pymilvus.exceptions.MilvusException: <MilvusException: (code=1, message=can't find collection name:book)>


```

milvus. log: 
```
[2023/06/06 02:59:12.429 +00:00] [WARN] [rootcoord/root_coord.go:1006] ["failed to describe collection"] [traceID=2bef844752b997d4] ["collection name"=book] [id=0] [msgID=441980119251943426] [ts=18446744073709551615] [allowUnavailable=false] [error="code: CollectionNotExists, reason: can't find collection: book"]
[2023/06/06 02:59:12.438 +00:00] [WARN] [retry/retry.go:44] ["retry func failed"] ["retry time"=0] [error="code: CollectionNotExists, reason: can't find collection: book"]
[2023/06/06 02:59:12.439 +00:00] [INFO] [proxy/impl.go:689] ["DescribeCollection done"] [traceID=2bef844752b997d4] [role=proxy] [MsgID=441980119251943426] [BeginTS=441980119251943426] [EndTS=441980119251943426] [db=] [collection=book]
```

**Comment by bigsheeper** [comment]
working on it

**Comment by bigsheeper** [comment]
Root cause:

![image](https://github.com/milvus-io/milvus/assets/42060877/0e2698af-dc45-4a57-b63e-5bf5f248ffbf)

Proxy should hold lock from `GetPartitionInfo` start to end.

**Comment by xiaofan-luan** [comment]
> Root cause:
> 
> ![image](https://user-images.githubusercontent.com/42060877/243900647-0e2698af-dc45-4a57-b63e-5bf5f248ffbf.png)
> 
> Proxy should hold lock from `GetPartitionInfo` start to end.

The implementation is weird.
Why do we get the collectionInfo at the very begining but still use the GetCollectionID

**Comment by xiaofan-luan** [comment]
on the other side,
we should use m.GetCollectionInfo() instead of m.collInfo()

**Comment by bigsheeper** [comment]
> on the other side, we should use m.GetCollectionInfo() instead of m.collInfo()

Totally aggreed

**Comment by bigsheeper** [comment]
Hello @AlexeyIvanov8 , thank you for your feedback. This issue has been resolved and shall be made available in the upcoming 2.3 version.

**Comment by bigsheeper** [comment]
Hello @AlexeyIvanov8 , thank you for your feedback. This issue has been resolved and shall be made available in the upcoming 2.3 version.
